### PR TITLE
Refactor to allow arbitrary form group layout components

### PR DESCRIPTION
**Pull request by Will-Howard**
_This will be against master once https://github.com/ForumMagnum/ForumMagnum/pull/8764 is merged_

This PR adds the ability to use arbitrary components for rendering form groups, defined in the formGroups "schema", similar to how `control: 'PostsListEditor'` works in the regular schema. You can tell a group to use a different layout component by setting the `layoutComponent` field:

```typescript
title: {
    name: "title",
    order: -10,
    layoutComponent: "MyCustomLayoutComponent"
    layoutComponentProps: {
      groupStyling: false,
      paddingStyling: true,
      flexAlignTopStyling: true
    }
  },
```

The motivations for this are that:
1. Sometimes you want a bit of UI in a form that doesn't exactly correspond to a field
2. Sometimes you want to style a form group in a fairly custom way (e.g. with `flex-direction: row`), and the current system of adding props to `FormGroupLayout` for different styling options isn't very scaleable

The more immediate motivation is that I want to use this in the [Google doc import PR](https://github.com/ForumMagnum/ForumMagnum/pull/8756), which has some fairly custom form group styling (see [designs](https://www.figma.com/file/IRUVQ5D2mKxoW0xZnhXIWv/2024-Q1?type=design&node-id=520-9331&mode=design&t=lNADovTuLmnDQn0u-0)). I have also run into both (1) and (2) recently with the tabbed edit post page redesign, and the social preview section.

Limitations of this PR:
- There aren't any examples of custom components being used yet
- In order to get a form group to display at all you still need to have _a_ visible field in it, to properly support (1) this shouldn't be the case. This wouldn't be that hard to support, I just decided to quit while I was ahead seeing as this isn't needed now (you would just need to add `formGroups` as a parameter when creating the collection and pass it through to `packages/lesswrong/components/vulcan-forms/Form.tsx`)

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1206629631662252) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/components/common/withDialog.tsx [hunk 1]
```diff
@@ -4,12 +4,10 @@ import { hookToHoc } from '../../lib/hocUtils';
 import { useTracking } from '../../lib/analyticsEvents';
 import { useOnNavigate } from '../hooks/useOnNavigate';
 
-export type CloseableComponents = {
-  [T in keyof ComponentTypes]: FromPartial<ComponentTypes[T]['propTypes']> extends { onClose: any } | undefined ? T : never
-}[keyof ComponentTypes];
+export type CloseableComponent = ComponentWithProps<{ onClose?: any }>
```
Comment (Will-Howard) [inline packages/lesswrong/components/common/withDialog.tsx:7]: I refactored this type definition into `ComponentWithProps`, which I also used to enforce the type of form group layout components. `ComponentWithProps` no longer uses the legacy [propTypes](https://legacy.reactjs.org/docs/typechecking-with-proptypes.html) property to check the type

File: packages/lesswrong/components/form-components/CoauthorsListEditor.tsx [hunk 2]
```diff
@@ -10,7 +10,7 @@ const coauthorsListEditorStyles = (theme: ThemeType): JssStyles => ({
   root: {
     display: 'flex',
     marginLeft: 8,
-    marginTop: -4,
+    marginTop: -12,
```
Comment (Will-Howard) [inline packages/lesswrong/components/form-components/CoauthorsListEditor.tsx:13]: This was the only form component that had `hideHeader` set to true, and there was some special styling based on this in FormGroupLayout to reduce the padding in this case. I've moved that to this file (with unfortunate negative margin) so avoid having to handle `hideHeader` in the layout component

File: packages/lesswrong/components/vulcan-forms/FormGroup.tsx [hunk 3]
```diff
@@ -47,212 +51,126 @@ const FormGroupHeaderComponent = registerComponent('FormGroupHeader', FormGroupH
...
-  render() {
-    const { name, fields, formComponents, label, defaultStyle, hideHeader, flexStyle, flexAlignTopStyle, paddingStyle, formProps } = this.props;
-    const { collapsed } = this.state;
-    const FormComponents = mergeWithComponents(formComponents);
-    const groupStyling = !(name === 'default' || defaultStyle)
-    const showHeading = groupStyling && !hideHeader
-
-    return (
-      <FormComponents.FormGroupLayout
+const FormGroup = ({
```
Comment (Will-Howard) [inline packages/lesswrong/components/vulcan-forms/FormGroup.tsx:74]: The changes in this file are:
- Refactoring `FormGroup` to be a function component
- Supporting LayoutComponents other than the default `FormGroupLayout`
- Some minor changes to what is passed through to `LayoutComponent`/`FormGroupLayout`

Comment (s-cheng) [reply packages/lesswrong/components/vulcan-forms/FormGroup.tsx:74]: 🙌

**Review by s-cheng — APPROVED** [review]
LGTM - appreciate all the comments, thanks!

**Pull request merged**